### PR TITLE
Add middleware config model

### PR DESCRIPTION
**Pull request by raymondfeng**
/to @ritch 

Connect to strongloop/loopback#1428

### CODE REVIEW

File: models/facet.js [hunk 1]
```diff
@@ -132,7 +134,15 @@ Facet.loadIntoCache = function(cache, facetName, allConfigFiles, cb) {
...
-      
+      cb();
+    });
+  }
+
+  if(middlewares) {
+    steps.push(function(cb) {
+      middlewares.load(cb);
+    }, function(cb) {
+      Middleware.deserialize(cache, facetName, middlewares);
```
Comment (bajtos) [inline models/facet.js]: IIRC, other entities use different method names instead of serialize/deserialize, it would be nice to preserve consistency in naming.

File: test/middleware.js [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+      });
+    });
+    it('should be persist multiple to the config file', function(done) {
+      var defs = Object.keys(this.configFile.data).sort();
+      expect(defs).to.eql(['auth', 'routes'].sort());
+      done();
+    });
+
+    it('should not contain workspace-private properties', function(done) {
+      // This test is reproducing an issue discovered in generator-loopback
```
Comment (bajtos) [inline test/middleware.js]: Is this comment true? Looks like a result of copy and paste to me.

File: models/facet.js (hunk 1 above)
Comment (raymondfeng) [inline models/facet.js]: Other entities have the logic embedded in facet.js, not as separate functions on the entity model itself.

File: models/middleware.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+      def = defs[d];
+      def.configFile = configFile.path;
+      def.phase = phase;
+      def.facetName = facetName;
+      def.name = d;
+      debug('loading [%s] middleware into cache', def.name);
+      Middleware.addToCache(cache, def);
+    }
+    if (!def) {
+      // Keep the phase information by adding an empty middleware config
```
Comment (bajtos) [inline models/middleware.js]: AFAICT, there is not unit-test to verify how this special entry is handled. Could you please add one?

File: models/middleware.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+      def = {
+        configFile: configFile.path,
+        phase: phase,
+        facetName: facetName,
+        name: '*' + phase // Set the name to be unique
```
Comment (bajtos) [inline models/middleware.js]: Is this unique enough? Can we use an internal property (a flag) instead?

File: models/facet.js (hunk 1 above)
Comment (bajtos) [inline models/facet.js]: I looked up the methods I was referring to: `ModelDefinition.getConfigFromCache`, `WorkspaceEntity.getConfigFromData`, `WorkspaceEntity.getDataFromConfig`. If I read the code in this patch correctly, your new serialize/deserialize methods call these `get*From*` method under the hood :+1: 

I agree it's a good idea to keep this logic out of facet.js.

Consider this comment as resolved then.

**Comment by bajtos** [comment]
Could you please add a test to verify that the order of phases is preserved after a load/save cycle?

Also what is the API for adding a new (empty) middleware phase? Could you please write a test for that too? *As a workspace API user, I'd like to define (insert) a new middleware phase at a given position.*

**Comment by ritch** [comment]
My biggest concern is breaking existing apps that use middleware. Writing a test that ensures order (according to lb runtime) is preserved for an app would ensure that won't happen.

**Comment by bajtos** [comment]
Few more scenarios (edge cases?) to consider:
 - What is the API call to reorder middleware phases? (Think of Arc GUI implementing middleware & phase editor.)
 - What happens when the last middleware from a phase is removed, either by deleting the middleware or by moving it to a different phase? IMO, the empty middleware phase should be preserved. Consider the following scenario: remove the default error handler, add a custom one.

**Comment by bajtos** [comment]
One more thing to consider: At the moment, project configuration that has a representation in Workspace API (Model, DataSource), is scaffolded via [component.js](https://github.com/strongloop/loopback-workspace/blob/05442a8754123a7fc77991f3c7a94b9d87470c26/templates/components/api-server/component.js) file. File templates that are copied as-is are used only for configuration that cannot be represented via Workspace API.

So far, there was no API for working with middleware, therefore the default middleware config is specified in a file [server/workspace.json](https://github.com/strongloop/loopback-workspace/blob/05442a8754123a7fc77991f3c7a94b9d87470c26/templates/components/api-server/template/server/middleware.json).

Now that we have API for middleware too, it is probably better to refactor `server/middleware.json` and move the template data to `component.js`, in order to preserve consistency.

**Comment by raymondfeng** [comment]
@bajtos Added the ability to keep the order of entries and related tests. PTAL.

**Comment by bajtos** [comment]
I think the proposed API will be rather difficult to use correctly (e.g. what is the API call to reorder middleware phases?), but if this is what works for your particular use case, then I am ok with it. I'll let @ritch to make a detailed review.

**Comment by raymondfeng** [comment]
We can refine the (not end-user facing) apis over time based on the use cases. Maybe I'll start to add `slc loopback:middleware` generator to exercise the apis.

Ideally, we should have the MiddlewarePhase and Middleware models. The Middleware belongsTo MiddlewarePhase. The MiddlewarePhase defines the phases and their ordering.

**Comment by bajtos** [comment]
> Maybe I'll start to add slc loopback:middleware generator to exercise the apis.

That would be great, as it would prove whether my concerns were relevant or not. I would personally like to see a consumer of this API implemented before we make a new release of loopback-workspace to npmjs.org, so that we don't have to bump up the major version if we find that this new API need breaking changes.

> Ideally, we should have the MiddlewarePhase and Middleware models. The Middleware belongsTo MiddlewarePhase. The MiddlewarePhase defines the phases and their ordering.

Sounds good to me :+1:

**Comment by bajtos** [comment]
@slnode test please

**Comment by bajtos** [comment]
@raymondfeng please consider extending the method `addPhase` with a check that the new phase is not inserted between sub-phases, i.e. a request to insert "audit" before "auth:after" should be rejected if there is "auth" or "auth:before". See https://github.com/strongloop/generator-loopback/pull/104#issuecomment-111213797 for more details.

**Comment by raymondfeng** [comment]
@bajtos The latest version supports sub-phases. PTAL.

File: models.json [hunk 5]
```diff
@@ -381,5 +381,69 @@
         }
       }
     }
+  },
+  "Middleware": {
+    "properties": {
+      "id": {"type": "string", "id": true, "json": false},
+      "name": {"type": "string", "json": false},
+      "phase": {"type": "string", "json": false},
```
Comment (bajtos) [inline models.json]: I think that `MiddlewarePhase` relation "hasMany phases" will add a property `Middleware.phaseId`. Is it a good idea to have both `phase` and `phaseId`? Should the model ensure that these two properties are always consistent, i.e. prevent a situation where `phase` points to a different entity than `phaseId`?

File: models/middleware.js [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+ * @param {Function} cb Callback function
+ */
+Middleware.getPhases = function(cb) {
+  // Load all entries from cache and sort them by order
+  Middleware.find(function(err, instances) {
+    if (err) return cb(err);
+    var phases = sortMiddleware(instances).phases;
+    cb(null, phases);
+  });
+}
```
Comment (bajtos) [inline models/middleware.js]: I think this method belongs to `MiddlewarePhase`, what do you think?

File: models/middleware.js [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+
+var ORDER_BUFFER = 1024;
+
+/**
+ * Add a middleware phase to the cache in a slot after the `nextPhase`
+ * @param {String} facetName
+ * @param {String} phase
+ * @param {String} nextPhase
+ */
+Middleware.addPhase = function(facetName, phase, nextPhase, cb) {
```
Comment (bajtos) [inline models/middleware.js]: Ditto, I think this method should be implemented by `MiddlewarePhase`. What are the benefits of keeping it here, in `Middleware`?

File: models.json [hunk 8]
```diff
@@ -381,5 +381,69 @@
...
+      "relations": {
+        "facet": {
+          "type": "belongsTo",
+          "model": "Facet",
+          "foreignKey": "facetName"
+        }
+      }
+    }
+  },
+  "MiddlewarePhase": {
```
Comment (bajtos) [inline models.json]: It looks like `MiddlewarePhase` is defined in JSON, but it is never created nor referenced anywhere. It looks like a dead code to me. We should either remove this model, or else put it to use. Thoughts?

File: models/middleware.js [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+ * @param {String} facetName Facet name
+ * @returns {ConfigFile}
+ */
+Middleware.serialize = function(cache, facetName) {
+  var middlewarePath = path.join(facetName, 'middleware.json');
+  var phases = loadFromCache(cache);
+  var middlewareConfig = {};
+
+  function addEntries(phase, subPhase, entries) {
+    if (Array.isArray(entries)) {
```
Comment (bajtos) [inline models/middleware.js]: Nitpick: use inverse condition + early return, it makes the code easier to understand and less nested.

```js
if (!Array.isArray(entries) return 0;
entries.forEach(function(m)  {
  // ...
});
return entries.length;
```

File: test/middleware.js [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+          nextPhase: 'routes'
+        }, done);
+      }], done);
+    });
+
+    beforeEach(function(done) {
+      this.configFile.load(done);
+    });
+
+    it('should be able to create multiple', function(done) {
```
Comment (bajtos) [inline test/middleware.js]: "it should be able to create multiple" - multiple what? phases? middleware entries? something else?

File: test/middleware.js [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+          if (err) done(err);
+          var middlewares = configFile.data;
+          expect(Object.keys(middlewares)).to.eql(
+            ['auth', 'routes:before', 'routes', 'myPhase']);
+          done();
+        });
+      });
+    });
+
+    it('should add phase after the first phase', function(done) {
```
Comment (bajtos) [inline test/middleware.js]: I think this adds the new phase at the end, i.e. the test should be called "...after the last phase".

File: test/middleware.js [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+          if (err) done(err);
+          var middlewares = configFile.data;
+          expect(Object.keys(middlewares)).to.eql(
+            ['auth', 'routes:before', 'routes', 'myPhase', 'phase1']);
+          done();
+        });
+      });
+    });
+  });
+  it('validates `name` uniqueness within the facet only', function(done) {
```
Comment (bajtos) [inline test/middleware.js]: Please add an empty line between tests to visually delimit them.

**Comment by bajtos** [comment]
@raymondfeng reviewed, see my comments above. The implementation looks good in general, I think the main question is what to do with `MiddlewarePhase` model. 

Please fix the failing CI build too.

File: models.json (hunk 8 above)
Comment (raymondfeng) [inline models.json]: I'll get rid of MiddlePhase for now.
